### DISCUSSION TITLE
When does Vercel think Next 13 will be production ready?

### DISCUSSION THREAD

**Discussion by james-beston** [vercel/next.js#44031] [Help]
### Summary

Hello!

I am about to dive into a pretty heavy project that needs to be deployed to at least a beta level by the start of April 2023. I am keen to employ the latest Next developments, so I was looking at building this out using the app folder in next13, but I am aware that this is not production ready.

Should I stick with 12 for now and look at porting to 13 further down the line?

### Additional information

The project is for a charity and would manage all of the signups, user engagement and playing audio for around 15k users.

**Comment by swarnava** [comment]
Next.js 13 is Stable and can be used for production application. There are some additional feature such as App directory, React Server Component which are currently in beta channel.

**Reply by james-beston** [reply]
Thanks @swarnava.
It is the App directory that I am mainly interested in. I am happy to build the production site using the App folder, but I don't want to fall foul of any issues. So you think I should continue with the Pages folder for now?

**Reply by swarnava** [reply]
Hi James, Yes that's correct. App directory is still in beta and we are polishing the feature currently. It is not meant to production use yet.

**Reply by clipartinc** [reply]
I tried using Next13 for a new project, but went back to Next12 for now. Using next-auth has changed a bit in Next13 and I need to learn the new ways. Looking forward to Next13 though as it grows.

**Comment by MiguelMarroquin** [comment]
Some updates about this topic? I think later will be a nightmare migrate to "APP"  ...

**Reply by asdillon** [reply]
I'm using Nextjs 13 on production project. We are still using the pages directory though.

**Reply by WassimBelhadjRhouma** [reply]
It is mentioned in the documentation that it is not recommended to use the app folder in production. We do not have an idea about when will we be able to use it. Instead, we can implement the nextjs 13's features in the page folder. The good news is that the pages folder will be supported in the future even after the release of the complete latest version.

**Reply by rolanday** [reply]
+1. It's great that Vercel is making appdir avail for preview, but it turned me away from the framework knowing every line of code I write in pages dir is incurring tech debt (knowing will want to change over to appdir, once stable, and for features already available in other platforms today). It's been 4 months since "launch" and no indication when Nextjs13 will be finished.

**Reply by Dev-Siri** [reply]
I am currently migrating a production app to the `app` directory. So far, the only drawbacks I faced were library support (many use context or browser APIs), the confusing architecture with Server & Client components, and TypeScript support (there aren't types like `NextPage` for the `app` directory, especially with the new file conventions).
But I haven't run into any breaking changes with the framework yet.

**Reply by TCGAUD** [reply]
I've been having issues with async server components.
Passing props to client components causes infinite render sometimes.

**Reply by shrekuu** [reply]
> Some updates about this topic? I think later will be a nightmare migrate to "APP" ...

Hahaha. You are right, so I am migrating my Next.js 12 app to this Next.js 13 app directory now, before my app grows way too big, for production.

**Comment by jaapaurelio** [comment]
Same boat here. A timeline would be good so we can decide if we should start with the app dir and avoid technical debt from the start.

**Reply by stakmur** [reply]
I can't even begin to negotiate deadlines for new project without some idea of timeline. Please help on this answer. Thanks.

**Reply by edwinvelez** [reply]
Agreed. A timeline would be very helpful.

**Comment by musashi-garami** [comment]
Thumbs up on this, also wouldn't mind a rough estimate of when App Dir is production ready, starting a project and really want to use App Dir.

**Reply by shrekuu** [reply]
Use it now, or you will end up with a migration nightmare. Hahaha~ I am having this migration nightmare now.

**Reply by sathlerhe** [reply]
I'm building a new application now and I don't want to migrate for app dir like in the next month i release the application, so I'm using app dir now. I hope to not have any problem in production

**Reply by notKvS** [reply]
> I'm building a new application now and I don't want to migrate for app dir like in the next month i release the application, so I'm using app dir now. I hope to not have any problem in production

Have you faced any problems with new app dir in production till now?

**Reply by sathlerhe** [reply]
Its still in development. Asap in production I'll talk if have any problems. But in dev I just had problems in async components, but nothing to stop the development or anything like that

**Reply by shrekuu** [reply]
@sathlerhe  What problem? Nginx config problem? guess it's:

```
    location / {
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;
        proxy_set_header X-NginX-Proxy true;
        proxy_pass http://localhost:3000;
        proxy_http_version 1.1;
        proxy_cache_bypass $http_upgrade;
        proxy_set_header Upgrade $http_upgrade;
        proxy_buffering off;   # Don't forget this line. This helps to flush any ready content immediately to the browser.
    }
```

**Reply by sathlerhe** [reply]
No, it was just not fetching data sometimes. I don't know what it was, but seems resolved.

**Reply by sathlerhe** [reply]
> > I'm building a new application now and I don't want to migrate for app dir like in the next month i release the application, so I'm using app dir now. I hope to not have any problem in production
> 
> Have you faced any problems with new app dir in production till now?

I released it to staging today, and sometimes routes doesn't work well. Besides it everything is alright, the pages loading is very very fast. I'm really excited with this new version and app dir. Just some errors to be solved, but almost everything is nice. I deployed it with dockerhub and kubernets.

**Comment by timneutkens** [comment]
https://nextjs.org/blog/next-13-3

> We are closing in on marking App Router as stable in the next minor release and are shifting our focus to optimizing performance, enhancing behaviors, and fixing bugs.

> While we're still working on a few features like Mutations we don't expect these to impact the API surface of other App Router features. We're excited to see what you build with the App Router and [look forward to your feedback](https://github.com/vercel/next.js/discussions/41745).

**Reply by roger-rangel** [reply]
Do you recommend we start using App Router and NextJs13 now? I just don't want to have a production-ready app with Nextjs12 and then have to migrate everything once again 😅 would be alot of technical debt and a migration nightmare for sure 👀

**Reply by fullstackjedi** [reply]
> Do you recommend we start using App Router and NextJs13 now? I just don't want to have a production-ready app with Nextjs12 and then have to migrate everything once again 😅 would be alot of technical debt and a migration nightmare for sure 👀

I have been using in prod for 2 months now with no issues.

**Reply by roger-rangel** [reply]
good to hear. I also think it may not cause issues as well. will wait a few weeks though for the next minor release ✅

**Reply by Galanthus** [reply]
Using the App folder results in extreme slow start up in development mode. Compiling pages in dev mode takes ages. Is there a fix?

**Reply by TCGAUD** [reply]
If you have a large application and using the package prettier.
Its going to be very slow. Because it goes through each and every file.
If you are using it, try remove it, then check if its still lagging @Galanthus

**Reply by Galanthus** [reply]
> If you have a large application and using the package prettier. Its going to be very slow. Because it goes through each and every file. If you are using it, try remove it, then check if its still lagging @Galanthus

Yes i am using Prettier check out my code under do you see something strange from the packages besides Prettier? 

Thank you for the fast reply. This is my package.json:

```
{
  "name": "ffb-next-headless-sanity",
  "author": "Nino",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "killport": "kill -9 $(lsof -ti:3000)",
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@heroicons/react": "^2.0.17",
    "@hookform/resolvers": "^3.1.0",
    "@portabletext/react": "^3.0.0",
    "@sanity/code-input": "^4.1.0",
    "@sanity/vision": "^3.9.1",
    "@tailwindcss/typography": "^0.5.9",
    "@types/node": "18.16.3",
    "@types/react": "18.2.0",
    "@types/react-dom": "18.2.1",
    "@types/validator": "^13.7.15",
    "@vercel/analytics": "^1.0.0",
    "autoprefixer": "^10.4.14",
    "clsx": "^1.2.1",
    "daisyui": "^2.51.6",
    "eslint": "8.39.0",
    "eslint-config-next": "^13.3.4",
    "framer-motion": "^10.12.5",
    "get-youtube-id": "^1.0.1",
    "next": "^13.3.4",
    "next-sanity": "^4.2.0",
    "next-themes": "^0.2.1",
    "postcss": "^8.4.23",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-hook-form": "^7.43.9",
    "react-icons": "^4.8.0",
    "react-lite-youtube-embed": "^2.3.52",
    "sanity": "^3.9.1",
    "sanity-plugin-asset-source-unsplash": "^1.0.6",
    "sanity-plugin-media": "^2.0.5",
    "sharp": "^0.32.1",
    "slugify": "^1.6.6",
    "styled-components": "^5.3.10",
    "swiper": "^9.2.4",
    "tailwindcss": "^3.3.2",
    "typescript": "5.0.4",
    "validator": "^13.9.0",
    "zod": "^3.21.4"
  }
}
```

tsconfig.json:

```
{
  "compilerOptions": {
    "target": "ES2017",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "baseUrl": ".",
    "paths": {
      "@/*": ["*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}
```

next.config.js:

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  poweredByHeader: false,
  reactStrictMode: true,
  experimental: {
    urlImports: ["https://themer.sanity.build/"],
    appDir: true
  },
  images: {
    domains: ["cdn.sanity.io", "images.pexels.com", "source.unsplash.com"]
  }
}

module.exports = nextConfig
```

**Reply by TCGAUD** [reply]
If you have a lot of CSS, it's worth noting that having all of it in a global file can potentially slow down your app's load times.
(IDK because i cant see your project)

,
Looking at your packages,
eslint, styled-components and Sanity might slow down your app's load times, especially if you've got a lot of content. If things are running slow, try disabling eslint and prettier to see if that helps speed things up.

if its still super slow.
You can also use a profiling tool to investigate whats taking so long. @Galanthus 
Goodluck man

**Reply by TCGAUD** [reply]
also just curious. lmao
How long does it take to start the dev server ?

**Reply by roger-rangel** [reply]
@TCGAUD do you think tailwind css + nextjs13 is a good mix? what css libraries are y'all using?
I deleted the prettier package since its not necessary

**Reply by Galanthus** [reply]
> If you have a lot of CSS, it's worth noting that having all of it in a global file can potentially slow down your app's load times. (IDK because i cant see your project)
> 
> , Looking at your packages, eslint, styled-components and Sanity might slow down your app's load times, especially if you've got a lot of content. If things are running slow, try disabling eslint and prettier to see if that helps speed things up.
> 
> if its still super slow. You can also use a profiling tool to investigate whats taking so long. @Galanthus Goodluck man

Thanks! I have almost no custom CSS it's all class based and coming directly from Tailwind. I do have UI kit:

const defaultTheme = require("tailwindcss/defaultTheme")

```
module.exports = {
  darkMode: "class",
  content: ["./app/**/*.{ts,tsx}", "./components/**/*.{ts,tsx}"],
  theme: {
    extend: {
      borderWidth: {
        1: "1px"
      },
      maxWidth: {
        "8xl": "1320px"
      },
      colors: {
        sun: "#F9B700",
        midnight: "#021228"
      },
      fontFamily: {
        sans: ["var(--poppins-font)", ...defaultTheme.fontFamily.sans],
        playfair: "var(--playfair-font)"
      }
    }
  },
  daisyui: {
    logs: false,
    themes: [
      {
        custom: {
          primary: "#F9B700",
          secondary: "#021228",
          accent: "#E07415",
          neutral: "#EEE",
          "base-100": "#ffffff",
          "--btn-focus-scale": 1,
          info: "#3ABFF8",
          success: "#36D399",
          warning: "#FBBD23",
          error: "#F87272"
        }
      }
    ]
  },
  plugins: [require("@tailwindcss/typography"), require("daisyui")]
}
```

I will try disabling from ur advice.

**Reply by Galanthus** [reply]
> also just curious. lmao How long does it take to start the dev server ?

After debugging with **@next/bundle-analyzer** came to the conclusion that reason why my app was extremely slow is the library **react-icons** the chunk of this node package is 7MB while the complete website is not even close to 3MB.

After removing this library the load time is 50% faster not from **25s** to **5s** startup in development mode and switching routes takes around 2 to 3 seconds.

**Comment by rolanday** [comment]
Back to topic, with all those new features Vercel announced today (blob storage, redis, postgres) , maybe this means they anounce in Thursday's keynote that they've finished what they announced 6 mo ago?

**Comment by dvakatsiienko** [comment]
IMPO not earlier than 6 months, because app router has a lot of unstable pieces here and there.
even including hot reloading issues, sudden subtle crashes during module resolution, etc.
Not to mention existing ecosystem integration into app router's design usage **in a right way** (meaning server components and component composition).

**Comment by Teshie** [comment]
@james-beston Now Next 13.4 is production ready just announced yesterday. Checkout here for more (https://nextjs.org/blog/next-13-4)
